### PR TITLE
Do not replace class but use a separate generated file instead

### PR DESCRIPTION
**Pull request by Toflar**
Currently, on every `composer install` or `composer update`, we actually dynamically replace the `PluginLoader` class with the defined plugins. While this might be fine for production use, it has two major drawbacks during development:

- Because the `PluginLoader` class is dynamically replaced, you'll end up having changes all the time. So if you include the `contao/manager-plugin` package via a `path` Composer repository and you want to try out your changes, `PluginLoader` always gets updated ending up in your change set.
- Composer will always complain because the class has changed (it executes a `git status --porcelain --untracked-files=no` and asks you if you want to discard the local changes)

Generally speaking, I also think it's bad practice to dynamically replace a PHP class that is then used.
This PR changes the logic so that the plugins are stored inside `vendor/contao/manager-plugin/generated/plugins.php` and included in the `PluginLoader`. It's still an immutable array so it can be cached by the OPCode cache.
The `/.generated` folder inside the directory is gitignored so `git status --porcelain --untracked-files=no` won't complain about changed files in there.

### CODE REVIEW

File: src/PluginLoader.php [hunk 1]
```diff
@@ -44,7 +40,12 @@ public function __construct(string $installedJson = null, array $plugins = null)
             @trigger_error('Passing the path to the Composer installed.json as first argument is no longer supported in version 2.3.', E_USER_DEPRECATED);
         }
 
-        $this->plugins = $plugins ?: [];
+        $this->plugins = null !== $plugins ? $plugins : array_filter((array) @include self::getGeneratedPath());
```
Comment (leofeyer) [inline src/PluginLoader.php]: Using `@include` is an anti-pattern, because it will also suppress error messages inside the included files.

**Comment by leofeyer** [comment]
> Generally speaking, I also think it's bad practice to dynamically replace a PHP class that is then used.

This is how **composer/package-versions-deprecated** works, too. I don‘t think it is bad practice per se.

But you should not commit the dynamic class into your Git repo, of course. Neither the `PackageVersion\Versions` class nor our `Contao\ManagerPlugin\PluginLoader` class.

File: src/PluginLoader.php (hunk 1 above)
Comment (Toflar) [reply src/PluginLoader.php]: These are generated files. We're talking performance here. Essentially it's the same as https://github.com/symfony/dependency-injection/blob/7cfcb0e2747aa4ce7ba92cde8f4d98b8dfce00a0/Dumper/PhpDumper.php#L588.

**Comment by Toflar** [comment]
Which is what we're doing at the moment though, that's why I've extracted it :D

File: src/PluginLoader.php (hunk 1 above)
Comment (leofeyer) [reply src/PluginLoader.php]: No, it isn‘t? The `include_once` call in this line is not silenced.

Comment (Toflar) [reply src/PluginLoader.php]: So how am I supposed to include without an additional file check?

Comment (leofeyer) [reply src/PluginLoader.php]: You are not. A file check will be necessary here.

Comment (m-vo) [reply src/PluginLoader.php]: IIRC using the error suppression variant is actually slower than a `file_exists` check in most cases. But I've got no source for that at hand.

Comment (m-vo) [reply src/PluginLoader.php]: Did a naive test out of interest. This is the time in µs for 1M iterations each (lower numbers are better):

```
    existing / @:      6.1252429485321
non-existing / @:      8.4501919746399
    existing / check:  8.349004983902
non-existing / check:  1.2823150157928
```

So seems it's marginally better to use `@` when the file exists but much faster with the check if it doesn't. But all in all the difference is extremely small and only shows when executed billions of times in a row.

Comment (Toflar) [reply src/PluginLoader.php]: Updated 👍

**Comment by aschempp** [comment]
Maybe we should follow `composer/package-versions-deprecated` and generate the class to `vendor/composer`?

**Comment by Toflar** [comment]
Is there any official recommendation as to where a generated PHP file (in order to improve performance) should be stored by any Composer plugin? In its own directory (`vendor/<vendor-name>/<package-name>`) or in `vendor/composer/<vendor-name>/<package-name>`? @Seldaek maybe 😊

**Comment by Seldaek** [comment]
composer/package-versions-deprecated generaets stuff in its own directory (much like ocramius/package-versions did. vendor/composer/InstalledVersions.php is generated by composer (2) itself not that plugin.

vendor/composer is as far as I am concerned reserved for our usage and anyone doing stuff in there better not complain if things break :p

**Comment by Seldaek** [comment]
After a quick glance at the diff, the PR seems sensible to me as it is 👍🏻

**Comment by Toflar** [comment]
This PR also fixes compat with Composer 2.3 :)

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**